False positive for W292

### How to reproduce:

```
cargo run resources/test/fixtures/C402.py
```

### Expected:

W292 is not reported. `resources/test/fixtures/C402.py` does end with a newline.

### Actual:

```
resources/test/fixtures/C402.py:1:35: W292 No newline at end of file
```

--- a/resources/test/fixtures/W292_2.py
+++ b/resources/test/fixtures/W292_2.py
@@ -1,0 +1,2 @@
+def fn() -> None:
+    pass

--- a/src/check_lines.rs
+++ b/src/check_lines.rs
@@ -114,41 +114,40 @@ pub fn check_lines(
     }
 
     // Enforce newlines at end of files.
-    if settings.enabled.contains(&CheckCode::W292) {
-        // If the file terminates with a newline, the last line should be an empty string slice.
+    if settings.enabled.contains(&CheckCode::W292) && !contents.ends_with('\n') {
+        // Note: if `lines.last()` is `None`, then `contents` is empty (and so we don't want to
+        // raise W292 anyway).
         if let Some(line) = lines.last() {
-            if !line.is_empty() {
-                let lineno = lines.len() - 1;
-                let noqa_lineno = noqa_line_for
-                    .get(lineno)
-                    .map(|lineno| lineno - 1)
-                    .unwrap_or(lineno);
+            let lineno = lines.len() - 1;
+            let noqa_lineno = noqa_line_for
+                .get(lineno)
+                .map(|lineno| lineno - 1)
+                .unwrap_or(lineno);
 
-                let noqa = noqa_directives
-                    .entry(noqa_lineno)
-                    .or_insert_with(|| (noqa::extract_noqa_directive(lines[noqa_lineno]), vec![]));
+            let noqa = noqa_directives
+                .entry(noqa_lineno)
+                .or_insert_with(|| (noqa::extract_noqa_directive(lines[noqa_lineno]), vec![]));
 
-                let check = Check::new(
-                    CheckKind::NoNewLineAtEndOfFile,
-                    Range {
-                        location: Location::new(lines.len(), line.len() + 1),
-                        end_location: Location::new(lines.len(), line.len() + 1),
-                    },
-                );
+            let check = Check::new(
+                CheckKind::NoNewLineAtEndOfFile,
+                Range {
+                    location: Location::new(lines.len(), line.len() + 1),
+                    end_location: Location::new(lines.len(), line.len() + 1),
+                },
+            );
 
-                match noqa {
-                    (Directive::All(_, _), matches) => {
-                        matches.push(check.kind.code().as_str());
-                    }
-                    (Directive::Codes(_, _, codes), matches) => {
-                        if codes.contains(&check.kind.code().as_str()) {
-                            matches.push(check.kind.code().as_str());
-                        } else {
-                            line_checks.push(check);
-                        }
-                    }
-                    (Directive::None, _) => line_checks.push(check),
+            match noqa {
+                (Directive::All(_, _), matches) => {
+                    matches.push(check.kind.code().as_str());
                 }
+                (Directive::Codes(_, _, codes), matches) => {
+                    if codes.contains(&check.kind.code().as_str()) {
+                        matches.push(check.kind.code().as_str());
+                    } else {
+                        line_checks.push(check);
+                    }
+                }
+                (Directive::None, _) => line_checks.push(check),
             }
         }
     }

--- a/src/linter.rs
+++ b/src/linter.rs
@@ -364,6 +364,18 @@ mod tests {
     }
 
     #[test]
+    fn w292_2() -> Result<()> {
+        let mut checks = check_path(
+            Path::new("./resources/test/fixtures/W292_2.py"),
+            &settings::Settings::for_rule(CheckCode::W292),
+            &fixer::Mode::Generate,
+        )?;
+        checks.sort_by_key(|check| check.location);
+        insta::assert_yaml_snapshot!(checks);
+        Ok(())
+    }
+
+    #[test]
     fn f401() -> Result<()> {
         let mut checks = check_path(
             Path::new("./resources/test/fixtures/F401.py"),

